libfreerdp/codec/yuv.c

```diff
@@ -290,48 +363,57 @@ static BOOL pool_decode(YUV_CONTEXT* context, PTP_WORK_CALLBACK cb, const BYTE*
 	}
 
 	/* case where we use threads */
-	steps = MAX((context->nthreads + numRegionRects / 2 + 1) / numRegionRects, 1);
-	nobjects = numRegionRects * steps;
-
-	if (!allocate_objects(&work_objects, (void**)&params, sizeof(YUV_PROCESS_WORK_PARAM), nobjects))
-		goto fail;
-
 	for (x = 0; x < numRegionRects; x++)
 	{
 		const RECTANGLE_16* rect = &regionRects[x];
-		const UINT32 height = rect->bottom - rect->top;
+		RECTANGLE_16 r = *rect;
 
-		const UINT32 heightStep = MAX((height + steps / 2 + 1) / steps, 1);
-		for (y = 0; y < steps; y++)
+		if (intersects(x, regionRects, numRegionRects))
+			continue;
+
+		while (r.left < r.right)
 		{
-			YUV_PROCESS_WORK_PARAM* cur = &params[waitCount];
-			RECTANGLE_16 r = *rect;
-			r.top += y * heightStep;
-
-			/* If we have an odd bounding rectangle we might end up with < steps
-			 * workers. Check we do not exceed the bounding rectangle. */
-			r.bottom = r.top + heightStep;
-			if (r.bottom > rect->bottom)
-				r.bottom = rect->bottom;
-			if (r.top >= rect->bottom)
-				continue;
-			if (r.bottom > yuvHeight)
-				r.bottom = yuvHeight;
-			*cur = pool_decode_param(&r, context, pYUVData, iStride, DstFormat, dest, nDstStep);
-			if (!submit_object(&work_objects[waitCount], cb, cur, context))
-				goto fail;
-			waitCount++;
+			RECTANGLE_16 y = r;
+			y.right = MIN(r.right, r.left + TILE_SIZE);
+
+			while (y.top < y.bottom)
+			{
+				RECTANGLE_16 z = y;
+				YUV_PROCESS_WORK_PARAM* cur;
+
+				if (context->work_object_count <= waitCount)
+				{
+					WLog_ERR(TAG, "YUV decoder: invalid number of tiles, only support %" PRIu32,
+					         context->work_object_count);
+					goto fail;
+				}
+
+				cur = &context->work_dec_params[waitCount];
+				z.bottom = MIN(z.bottom, z.top + TILE_SIZE);
+				if (rectangle_is_empty(&z))
+					continue;
+				*cur = pool_decode_param(&z, context, pYUVData, iStride, DstFormat, dest, nDstStep);
+				if (!submit_object(&context->work_objects[waitCount], cb, cur, context))
+					goto fail;
+				waitCount++;
+				y.top += TILE_SIZE;
+			}
+
+			r.left += TILE_SIZE;
 		}
 	}
 	rc = TRUE;
 fail:
-	free_objects(work_objects, params, nobjects);
+	free_objects(context->work_objects, waitCount);
```

Comment:
Doesn't this result in a dangling `work_objects` pointer in the context after the first `pool_decode` resulting in freed memory access when `pool_decode` is called again (unless the yuv context has been reset meanwhile)?

Reply:
no, it just cleans up the work objects internals, not the memory storing the struct